File: tensorflow/java/src/test/java/org/tensorflow/TensorMatcher.java
/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package org.tensorflow;

import org.hamcrest.TypeSafeMatcher;
import org.hamcrest.Description;

public class TensorMatcher {

  /**
  * Creates a matcher for (@link Tensor}s that only matches when the examined {@link Tensor} 
  * has a {@link DataType} that is equal to the specified <code>type</code>.
  * <p/>
  * For example:
  * <pre>assertThat(tensor, hasDataType(DataType.INT64);</pre>
  *
  * @param type 
  *     the type to compare against the DataType of the examined {@link Tensor} 
  */
  public static TypeSafeMatcher<Tensor> hasDataType(final DataType type) {
    return new TypeSafeMatcher<Tensor>() {
      @Override
      protected boolean matchesSafely(final Tensor item) {
        return type == item.dataType();
      }

      @Override
      public void describeTo(final Description description) {
        description.appendText("Tensor with data type of ").appendValue(type);
      }

      @Override
      public void describeMismatchSafely(final Tensor item, final Description description) {
        description.appendText(" was ");
        description.appendText("Tensor with data type of ").appendValue(item.dataType());
      }
    };
  }

  /**
  * Creates a matcher for {@link Tensor}s that only matches when the examined {@link Tensor}
  * has a {@link Tensor#numDimensions()} of zero, <strong>and</strong> a {@link Tensor#shape()}
  * length of zero.
  * <p/>
  * For example:
  * <pre>assertThat(tensor, is(scalar()));</pre>
  */
  public static TypeSafeMatcher<Tensor> scalar() {
    return new TypeSafeMatcher<Tensor>() {

      @Override
      protected boolean matchesSafely(final Tensor item) {
        final boolean hasScalarNumDimensions = 0 == item.numDimensions();
        final boolean hasScalarShape = 0 == item.shape().length;

        return hasScalarNumDimensions && hasScalarShape;
      }

      @Override
      public void describeTo(final Description description) {
        description.appendText("Tensor with shape length ").appendValue(0);
        description.appendText(" and number of dimensions ").appendValue(0);
      }

      @Override
      protected void describeMismatchSafely(final Tensor item, final Description description) {
        description.appendText(" was ");
        description.appendText("Tensor with shape length ").appendValue(item.shape().length);
        description.appendText(" and number of dimensions ").appendValue(item.numDimensions());
      }
    };
  }
}

File: tensorflow/core/framework/tensor_matcher.cc
/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
#include "tensorflow/core/framework/tensor_matcher.h"

#include <stdint.h>

#include <complex>
#include <ostream>
#include <string>

#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include "absl/log/log.h"
#include "absl/types/span.h"
#include "Eigen/Core"  // from @eigen_archive
#include "tensorflow/core/framework/numeric_types.h"
#include "tensorflow/core/framework/register_types.h"
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/framework/tensor_shape.h"
#include "tensorflow/core/framework/tensor_types.h"
#include "tensorflow/core/framework/types.h"
#include "tensorflow/core/platform/bfloat16.h"
#include "tensorflow/core/platform/tstring.h"
#include "tensorflow/core/platform/types.h"

namespace tensorflow {
namespace test {
namespace {

using tensorflow::Tensor;

template <typename T>
bool MatchAndExplainPointwise(absl::Span<const T> value,
                              absl::Span<const T> target,
                              ::testing::MatchResultListener* listener) {
  auto matcher = ::testing::MatcherCast<absl::Span<const T>>(
      ::testing::Pointwise(::testing::Eq(), target));
  return matcher.MatchAndExplain(value, listener);
}

class TensorEqMatcherImpl : public ::testing::MatcherInterface<const Tensor&> {
 public:
  explicit TensorEqMatcherImpl(const Tensor& target) : target_(target) {}

  void DescribeTo(::std::ostream* os) const override {
    *os << "data type is " << tensorflow::DataTypeString(target_.dtype())
        << ", and shape is " << target_.shape();
    switch (target_.dtype()) {
#define CASE_TYPE(T)                                       \
  case tensorflow::DataTypeToEnum<T>::value: {             \
    *os << ", and tensor data ";                           \
    absl::Span<const T> data(target_.unaligned_flat<T>()); \
    ::testing::MatcherCast<absl::Span<const T>>(           \
        ::testing::Pointwise(::testing::Eq(), data))       \
        .DescribeTo(os);                                   \
    break;                                                 \
  }
      TF_CALL_POD_STRING_TYPES(CASE_TYPE);
#undef CASE_TYPE
      default: {
        DLOG(FATAL) << "TensorEq matcher unsupported dtype: "
                    << tensorflow::DataTypeString(target_.dtype());
      }
    }
  }

  void DescribeNegationTo(::std::ostream* os) const override {
    *os << "data type is not " << tensorflow::DataTypeString(target_.dtype())
        << ", or shape is not " << target_.shape();
    switch (target_.dtype()) {
#define CASE_TYPE(T)                                       \
  case tensorflow::DataTypeToEnum<T>::value: {             \
    *os << ", or tensor data ";                            \
    absl::Span<const T> data(target_.unaligned_flat<T>()); \
    ::testing::MatcherCast<absl::Span<const T>>(           \
        ::testing::Pointwise(::testing::Eq(), data))       \
        .DescribeNegationTo(os);                           \
    break;                                                 \
  }
      TF_CALL_POD_STRING_TYPES(CASE_TYPE);
#undef CASE_TYPE
      default: {
        DLOG(FATAL) << "TensorEq matcher unsupported dtype: "
                    << tensorflow::DataTypeString(target_.dtype());
      }
    }
  }

  bool MatchAndExplain(
      const Tensor& value,
      ::testing::MatchResultListener* listener) const override {
    const bool dtype_compare = value.dtype() == target_.dtype();
    *listener << "whose data type " << tensorflow::DataTypeString(value.dtype())
              << (dtype_compare ? " matches " : " doesn't match ")
              << tensorflow::DataTypeString(target_.dtype());

    const bool shape_compare = value.shape() == target_.shape();
    *listener << ", whose shape " << value.shape()
              << (shape_compare ? " matches " : " doesn't match ")
              << target_.shape();

    if (!dtype_compare || !shape_compare) {
      return false;
    }

    // For POD-types, Tensor comparison can be done by comparing buffer returned
    // by tensor_data() functions. However, that does not give useful debug
    // information when match fails. Therefore we switch on data type.
    bool result;
    switch (target_.dtype()) {
#define CASE_TYPE(T)                                                       \
  case tensorflow::DataTypeToEnum<T>::value: {                             \
    result = MatchAndExplainPointwise<T>(                                  \
        value.unaligned_flat<T>(), target_.unaligned_flat<T>(), listener); \
    break;                                                                 \
  }
      TF_CALL_POD_STRING_TYPES(CASE_TYPE);
#undef CASE_TYPE
      default: {
        DLOG(FATAL) << "TensorEq matcher unsupported dtype: "
                    << tensorflow::DataTypeString(target_.dtype());
        result = false;
      }
    }

    return result;
  }

 private:
  const Tensor target_;
};

}  // namespace

TensorEq::operator ::testing::Matcher<const Tensor&>() const {
  return ::testing::MakeMatcher(new TensorEqMatcherImpl(target_));
}

}  // namespace test
}  // namespace tensorflow
